### ISSUE TITLE
Support for Sparse Group Lasso?

### ISSUE THREAD

**Issue by AlejandroCatalina** [scikit-learn-contrib/lightning#104]
Hi,

I've just found this library and seems pretty good, congratulations! I see that you support group lasso and I was wondering if there was a plan for supporting sparse group lasso, since that shouldn't be much of a trouble (with all due respect). I didn't dig in the code too much so it may even be there but, at first sight, I didn't see it.

I actually saw the @fabianp implementation of sparse group lasso in a personal repo from which you can have the standard group lasso solution just by setting the parameter alpha = 0 and I thought that it may be useful to have a fast and good implementation here.

Either way, thank you for this great work!

**Comment by AlejandroCatalina** [comment]
That's great, I'll take a deeper look at the code and see if I can give it
a go.

Thank you for the fast answer.

**Comment by AlejandroCatalina** [comment]
Could you point me to some documentation or algorithm to follow the code? I've been looking into it (specifically the solve_l1l2 function) and I don't really get it, i was expecting something more like https://github.com/fabianp/group_lasso/blob/master/group_lasso.py but doesn't seem so. I think it's harder to follow because it's a generic solver for any objective function, isn't it?

**Comment by fabianp** [comment]
Yes, in lightning there are multiple generic solvers. For sparse group lasso, you probably want to use [coordinate descent](https://github.com/scikit-learn-contrib/lightning/blob/master/lightning/impl/primal_cd.py) or FISTA. So implementing the sparse group lasso amount to define the appropriate prox or update rule for these methods.

**Comment by AlejandroCatalina** [comment]
Okay thanks. Is there any difference in terms of performance between FISTA and coordinate descent? In the implementation, I mean, I see that coordinate descent is implemented using Cython while FISTA only in Python.

I'd say that both are pretty competitive, but I'm interested in large applications, will FISTA hold against the current coordinate descent implementation?

**Comment by fabianp** [comment]
Simplifying a lot I would say that they both have similar performance, but this will depend depend on things like sparsity/number of features / strong convexity

**Comment by AlejandroCatalina** [comment]
Okay great. I've implemented FISTA but I would like to give a shot to coordinate descent. I'll pr when I get home.

Thank you for your time.

**Comment by fabianp** [comment]
For now we don't as the existing group lasso implementation considers that groups are equal to the coefficient associated with the different classes (is dependent of a multiclass formulation). I would add a parameter groups=[] to the class, of size n_features, where each entry specifies to which group the coefficient belongs to.

**Comment by AlejandroCatalina** [comment]
Yes that's what I thought. That would make the class useful for regression
wouldn't it? Right now there isn't a proper group lasso for regression, if
I didn't misunderstood you.

**Comment by fabianp** [comment]
You are right, there isn't

**Comment by AlejandroCatalina** [comment]
Is there a plan to include it? I am playing with it but I'm afraid I may implement it in a non-optimal way, which would hurt performance on large scale applications.

Sorry for the persistence.

**Comment by fabianp** [comment]
no plan for me to work on that in the short term.
